xapian-core/backends/glass/glass_version.cc

```diff
@@ -295,7 +295,20 @@ GlassVersion::write(glass_revision_number_t new_rev, int flags)
 	else
 	    tmpfile += "/v.tmp";
 
-	fd = posixy_open(tmpfile.c_str(), O_CREAT|O_TRUNC|O_WRONLY|O_BINARY, 0666);
+#ifdef __EMSCRIPTEN__
+        /*
+        * Emscripten crashes if using O_TRUNC here,
+        * so unlink and create a new file instead
+        */
+		unlink(tmpfile.c_str());
+		fd = posixy_open(tmpfile.c_str(),
+				O_CREAT|O_WRONLY|O_BINARY, 0666);
+#else
+		fd = posixy_open(tmpfile.c_str(),
+				O_CREAT|O_TRUNC|O_WRONLY|O_BINARY,
+				0666);
+#endif
```

Comment:
Emscripten crashes if trying to truncate a non existing file ( `v.tmp` in this case). It crashes on this line here in the emscripten filesystem sources: https://github.com/emscripten-core/emscripten/blob/incoming/src/library_fs.js#L1043

Reply:
I think there must be something more to this, since `backends/flint_lock.cc` opens the lock file with the `O_CREAT|O_TRUNC` flag combination, and that won't exist for a new DB:

```
#elif defined FLINTLOCK_USE_FLOCK
    // This is much simpler than using fcntl() due to saner semantics around
    // releasing locks when closing other descriptors on the same file (at
    // least on platforms where flock() isn't just a compatibility wrapper
    // around fcntl()).  We can't simply switch to this without breaking
    // locking compatibility with previous releases, though it might be useful 
    // for porting to platforms without fcntl() locking.  Also, flock()    
    // apparently doesn't work over NFS - perhaps that's OK, but we should at
    // least check the failure mode.
    Assert(fd == -1);
    int lockfd = open(filename.c_str(), O_WRONLY | O_CREAT | O_TRUNC | O_CLOEXEC, 0666);
```

And it looks to me like `io_open_block_wr()` in `common/io_utils.cc` is also affected when creating a database:

```
int
io_open_block_wr(const char * fname, bool anew)
{
    // Use auto because on AIX O_CLOEXEC may be a 64-bit integer constant.
    auto flags = O_RDWR | O_BINARY | O_CLOEXEC;
    if (anew) flags |= O_CREAT | O_TRUNC;
    int fd = ::open(fname, flags, 0666);
```

Both of those bits of code should also be getting executed.

I tried disabling this hunk and adding some code to report what file we try to open and what the result is, and that suggests that Emscripten actually fails to open the file, returning `-1` and then the xapian code throws an exception (which causes an `abort()` because that's how emscripten  handles exceptions by default).  But this only happens the **second time** this code gets executed and both times the file doesn't exists beforehand (the "exists" value is from checking the filename with `stat()`).

```
Opening tmp version file 'testdb/v.tmp': exists = 0
posixy_open() returned 4
Created database with 1 documents
Opening tmp version file 'testdb/v.tmp': exists = 0
posixy_open() returned -1 errno = No such file or directory
exception thrown: 5407248
```

Were you seeing something different?

This makes me wonder if the "create file with O_TRUNC", "rename it", "create file with O_TRUNC" sequence on the same filename is the key to this problem.

Reply:
Hacked deeper into emscripten and got an error message from where it happens:

```
{ Error: ENOENT: no such file or directory, utime 'testdb/v.tmp'
    at Object.fs.utimesSync (fs.js:1214:11)
```

Which is exactly here:

https://github.com/emscripten-core/emscripten/blob/incoming/src/library_nodefs.js#L138

If we go back to the `open` method in the emscripten filesystem API we find the truncate call here:

https://github.com/emscripten-core/emscripten/blob/incoming/src/library_fs.js#L1050

Looking into the truncate method we see the call to setattr:

https://github.com/emscripten-core/emscripten/blob/incoming/src/library_fs.js#L967

So the bug is clearly in emscripten, trying to set attributes on files that have not been created yet.

Reply:
> Another "workaround" could be creating the database folder before creating the `v.tmp` file.

That's what's already meant to happen, and testing `xapianjstest.cc` with the `EM_ASM_` section removed and compiled for Linux, that's indeed what happens.  If I add a `system` call to `ls -l` the directory before it tries to create `v.tmp` then it's there and the `flintlock` file has been created too.

So I'm confused - is that not what happens under emscripten?  You seem to be suggesting that we're trying to open `testdb/v.tmp` at a point before `testdb` has even been created.  I would expect the `open` to just fail in that case (and set `errno` to `ENOTDIR`).

It occurs to that the [handling for `O_TRUNC` here](https://github.com/emscripten-core/emscripten/blob/incoming/src/library_fs.js#L1049) could be moved inside the `if (!created) {`...`}` check that's right before it, since if we just created the file then it'll be zero size so there's no need to truncate it.

But even if that stopped this problem from manifesting, I'm not sure it's really fixing whatever is actually wrong here.

Reply:
You're right. You are creating the database directory before the `v.tmp` file. Me not doing my homework properly here.

So I've been digging again‚ Xapian code this time, and found that v.tmp is created again on increasing the version number when committing the transaction, and this is where the error happens. So what I see from the first time the `v.tmp` file is created, after writing and closing it it's renamed to `iamglass`.

Then when committing the transaction `v.tmp` is created again, and fails with emscripten.

So I think now that I've been able to reconstruct a similar scenario with this snippet:

```
int fd = open("test.txt",
			 O_CREAT|O_TRUNC|O_WRONLY,
			 0666);
    printf("FD %d\n", fd);
    
    write(fd, "test", 4);
    close(fd);
    
    rename("test.txt", "newtest.txt");
    
    fd = open("test.txt",
			 O_CREAT|O_TRUNC|O_WRONLY,
			 0666);
    printf("FD %d\n", fd);
```

Where the latest output is `FD -1`. The `rename` makes the difference here. So changing my theory to that emscripten has a problem with creating files with O_TRUNC when a file with the same name existed earlier, but was renamed.

Sounds like a better theory?

Reply:
Yes, that seems to fit all the facts, and in particular why only this call to `open()` seems to be affected.

Writing to a temporary file and renaming it is quite a common pattern, but I could believe it's unusual to repeat the pattern with the exact same temporary filename and within the same run of the program, so I could believe we were the first to actually hit this.

Could you open an issue against emscripten?

Reply:
Yes - I'll build a test case and issue for emscripten then. I suggest we keep the current workaround until this is resolved in emscripten.

Reply:
Sure - the workaround is fine for now (especially as we now understand why this particular call needs it and other cases don't).

Let me know when there's a ticket to link to and I'll update the comment for the workaround.

Reply:
@ojwb Issue created here: https://github.com/emscripten-core/emscripten/issues/8187

Reply:
Thanks, updated comment in 5e462cf60f3c373b8be4ac1411e409504bc452bd.